lifecycle-operator/controllers/common/schedulinggateshandler.go

```diff
@@ -0,0 +1,110 @@
+package common
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	klcv1alpha3 "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3"
+	apicommon "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3/common"
+	controllererrors "github.com/keptn/lifecycle-toolkit/lifecycle-operator/controllers/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate moq -pkg fake -skip-ensure -out ./fake/schedulinggateshandler_mock.go . ISchedulingGatesHandler
+type ISchedulingGatesHandler interface {
+	RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error
+	IsSchedullingEnabled() bool
+}
+
+type RemoveGatesFunc func(ctx context.Context, c client.Client, log logr.Logger, podName string, podNamespace string) error
+type GetPodsFunc func(ctx context.Context, c client.Client, log logr.Logger, ownerUID types.UID, ownerKind string, namespace string) ([]string, error)
+
+type SchedulingGatesHandler struct {
+	client.Client
+	logr.Logger
+	enabled     bool
+	removeGates RemoveGatesFunc
+	getPods     GetPodsFunc
+}
+
+func NewSchedulingGatesHandler(c client.Client, l logr.Logger, enabled bool) *SchedulingGatesHandler {
+	return &SchedulingGatesHandler{
+		Client:      c,
+		Logger:      l,
+		enabled:     enabled,
+		removeGates: removePodGates,
+		getPods:     getPodsOfOwner,
+	}
+}
+
+func (h *SchedulingGatesHandler) RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error {
+	switch workloadInstance.Spec.ResourceReference.Kind {
+	case "Pod":
+		return h.removeGates(ctx, h.Client, h.Logger, workloadInstance.Spec.ResourceReference.Name, workloadInstance.Namespace)
+	case "ReplicaSet", "StatefulSet", "DaemonSet":
+		podList, err := h.getPods(ctx, h.Client, h.Logger, workloadInstance.Spec.ResourceReference.UID, workloadInstance.Spec.ResourceReference.Kind, workloadInstance.Namespace)
```

Comment:
Why do we loop twice over the pod list here? Can't we do this in a single step? 

listpod in namespace
if owner == owner id remove gate

Reply:
easier to test when the retrieval of the pods and afterwards the gate removal is done in 2 steps

Reply:
otherwise you are not able to cover all cases if you put the functionality into a single function

Reply:
I am not sure I understand the difference in simplicity here, in what is now the getPods function you could call h.removeGates this is still mockable, you could call it GetAndRemove or similar and you get the benefit that in a test where only the get and removed is called you do not need to mock also the remove function